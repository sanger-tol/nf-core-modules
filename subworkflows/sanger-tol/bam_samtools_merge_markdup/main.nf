include { SAMTOOLS_FAIDX   } from '../../../modules/nf-core/samtools/faidx/main'
include { SAMTOOLS_MERGE   } from '../../../modules/nf-core/samtools/merge/main'
include { SAMTOOLS_MARKDUP } from '../../../modules/nf-core/samtools/markdup/main'

workflow BAM_SAMTOOLS_MERGE_MARKDUP {

    take:
    ch_bam              // channel: [ val(meta), [ bam ] ]
    ch_assemblies       // channel: [ val(meta), fasta ]
    val_mark_duplicates // boolean: mark duplicates on output bam

    main:
    ch_versions = Channel.empty()

    //
    // Module: Index assembly fastas
    //
    SAMTOOLS_FAIDX(
        ch_assemblies, // reference
        [ [:],[] ],    // fai
        false          // get sizes
    )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX.out.versions)

    //
    // Logic: create a channel with both fai and gzi for each assembly
    //        We do it here so we don't cause downstream issues with the
    //        remainder join
    //
    ch_fai_gzi = SAMTOOLS_FAIDX.out.fai
        | join(SAMTOOLS_FAIDX.out.gzi, by: 0, remainder: true)
        | map { meta, fai, gzi -> [ meta, fai, gzi ?: [] ] }

    //
    // Logic: Prepare input for merging bams.
    //        We use the ch_chunk_counts to set a groupKey so that
    //        we emit groups downstream ASAP once all bams have been made
    //
    ch_samtools_merge_input = ch_bam
        | combine(ch_assemblies, by: 0)
        | combine(ch_fai_gzi, by: 0)
        | multiMap { meta, bams, assembly, fai, gzi ->
            bam:   [ meta, bams ]
            fasta: [ meta, assembly ]
            fai:   [ meta, fai ]
            gzi:   [ meta, gzi ]
        }

    //
    // Module: Merge position-sorted bam files
    //
    SAMTOOLS_MERGE(
        ch_samtools_merge_input.bam,
        ch_samtools_merge_input.fasta,
        ch_samtools_merge_input.fai,
        ch_samtools_merge_input.gzi,
    )
    ch_versions = ch_versions.mix(SAMTOOLS_MERGE.out.versions)

    //
    // Module: Mark duplicates on the merged bam
    //
    ch_samtools_markdup_input = SAMTOOLS_MERGE.out.bam
        | combine(ch_assemblies, by: 0)
        | filter { val_mark_duplicates }
        | multiMap { meta, bam, assembly ->
            bam:      [ meta, bam ]
            assembly: [ meta, assembly ]
        }

    SAMTOOLS_MARKDUP(
        ch_samtools_markdup_input.bam,
        ch_samtools_markdup_input.assembly
    )
    ch_versions = ch_versions.mix(SAMTOOLS_MARKDUP.out.versions)

    emit:
    bam      = val_mark_duplicates ? SAMTOOLS_MARKDUP.out.bam : SAMTOOLS_MERGE.out.bam // channel: [ val(meta), [ bam ] ]
    versions = ch_versions                                                             // channel: [ versions.yml ]
}
