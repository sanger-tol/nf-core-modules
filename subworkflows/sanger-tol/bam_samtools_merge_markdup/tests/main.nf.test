nextflow_workflow {

    name "Test Subworkflow BAM_SAMTOOLS_MERGE_MARKDUP"
    script "../main.nf"
    workflow "BAM_SAMTOOLS_MERGE_MARKDUP"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/bam_samtools_merge_markdup"
    tag "samtools"
    tag "subworkflows/../../modules/nf-core/samtools/faidx"
    tag "subworkflows/../../modules/nf-core/samtools/merge"
    tag "subworkflows/../../modules/nf-core/samtools/markdup"

    config "./nextflow.config"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "samtools/merge",
                "samtools/markdup",
                "samtools/faidx"
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
    }

    test("Undibacterium unclassified - hap1 - no markdup") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/hic_mapping/41741_2.7.sub.cram.baUndUnlc1_hic_phased_hap1.2.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/hic_mapping/41741_2.7.sub.cram.baUndUnlc1_hic_phased_hap1.3.bam', checkIfExists: true),
                    ]
                ])
                input[1] = Channel.of([
                    [ id:'test' ],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.unscaffolded.fa.gz', checkIfExists: true),
                ])
                input[2] = false
                """
            }
        }

        then {

            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out.bam.collect { bam(it.get(1)).getStatistics() },
                    workflow.out.versions
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }

    }

    test("Undibacterium unclassified - hap1 - markdup") {

        when {

            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/hic_mapping/41741_2.7.sub.cram.baUndUnlc1_hic_phased_hap1.2.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'Undibacterium_unclassified/analysis/hic_mapping/41741_2.7.sub.cram.baUndUnlc1_hic_phased_hap1.3.bam', checkIfExists: true),
                    ]
                ])
                input[1] = Channel.of([
                    [ id:'test' ],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.unscaffolded.fa.gz', checkIfExists: true),
                ])
                input[2] = true
                """
            }
        }

        then {
            assert workflow.success
            assertAll(
                { assert snapshot(
                    workflow.out.bam.collect { bam(it.get(1)).getStatistics() },
                    workflow.out.versions
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }

    }
}
