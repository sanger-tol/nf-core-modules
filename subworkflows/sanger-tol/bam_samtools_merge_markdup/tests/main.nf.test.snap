{
    "Undibacterium unclassified - hap1 - no markdup": {
        "content": [
            [
                {
                    "maxReadLength": 151,
                    "minReadLength": 151,
                    "meanReadLength": 151,
                    "maxQuality": 60,
                    "minQuality": 10,
                    "meanQuality": 50,
                    "readCount": 11583,
                    "duplicateReadCount": 0,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,3810f6c8d241f6e18c118c3eee06c931",
                "versions.yml:md5,747adb74c8d8e46aeeb149c9053718b5"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-21T14:30:16.602926"
    },
    "Undibacterium unclassified - hap1 - markdup": {
        "content": [
            [
                {
                    "maxReadLength": 151,
                    "minReadLength": 151,
                    "meanReadLength": 151,
                    "maxQuality": 60,
                    "minQuality": 10,
                    "meanQuality": 50,
                    "readCount": 11583,
                    "duplicateReadCount": 508,
                    "sorted": false
                }
            ],
            [
                "versions.yml:md5,3810f6c8d241f6e18c118c3eee06c931",
                "versions.yml:md5,747adb74c8d8e46aeeb149c9053718b5",
                "versions.yml:md5,9c798a363e3c7297bebbaa008aefe259"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-21T14:33:09.850931"
    }
}